File: 09_intersectionPoint/9.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

int findLength(struct Node* head) {
    int i = 0;
    struct Node* temp = head;
    while (temp) {
        i++;
        temp = temp->next;
    }
    return i;
}

struct Node* intersectPoint(struct Node* head1, struct Node* head2) {
    int len1 = findLength(head1);
    int len2 = findLength(head2);

    struct Node* higher = len1 > len2 ? head1 : head2;
    struct Node* lower = len1 > len2 ? head2 : head1;

    for (int i = 0; i < abs(len1 - len2); i++) {
        higher = higher->next;
    }

    while (higher && lower) {
        if (higher == lower)
            return higher;
        higher = higher->next;
        lower = lower->next;
    }

    return NULL;
}

void printList(struct Node* head) {
    while (head) {
        printf("%d -> ", head->data);
        head = head->next;
    }
    printf("None\n");
}

int main() {
    struct Node* intersect = (struct Node*)malloc(sizeof(struct Node));
    intersect->data = 8;
    intersect->next = (struct Node*)malloc(sizeof(struct Node));
    intersect->next->data = 10;
    intersect->next->next = NULL;

    struct Node* head1 = (struct Node*)malloc(sizeof(struct Node));
    head1->data = 3;
    head1->next = (struct Node*)malloc(sizeof(struct Node));
    head1->next->data = 6;
    head1->next->next = (struct Node*)malloc(sizeof(struct Node));
    head1->next->next->data = 9;
    head1->next->next->next = intersect;

    struct Node* head2 = (struct Node*)malloc(sizeof(struct Node));
    head2->data = 4;
    head2->next = intersect;

    printf("List 1:\n");
    printList(head1);
    printf("List 2:\n");
    printList(head2);

    struct Node* result = intersectPoint(head1, head2);

    if (result)
        printf("Intersection at node with data: %d\n", result->data);
    else
        printf("No intersection found.\n");

    return 0;
}

File: 09_intersectionPoint/9.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Solution:
    def findLength(self, head):
        i = 0
        temp = head
        while temp:
            i += 1
            temp = temp.next
        return i

    def intersectPoint(self, head1, head2):
        len1 = self.findLength(head1)
        len2 = self.findLength(head2)

        higher = head1 if len1 > len2 else head2
        lower = head2 if len1 > len2 else head1

        for _ in range(abs(len1 - len2)):
            higher = higher.next

        while higher and lower:
            if higher == lower:
                return higher
            higher = higher.next
            lower = lower.next

        return None

def printList(head):
    while head:
        print(head.data, end=" -> ")
        head = head.next
    print("None")

intersect = Node(8)
intersect.next = Node(10)

head1 = Node(3)
head1.next = Node(6)
head1.next.next = Node(9)
head1.next.next.next = intersect

head2 = Node(4)
head2.next = intersect

print("List 1:")
printList(head1)
print("List 2:")
printList(head2)

sol = Solution()
result = sol.intersectPoint(head1, head2)

if result:
    print(f"Intersection at node with data: {result.data}")
else:
    print("No intersection found.")